apps/rush-lib/src/cli/utilities/pnpm/PnpmShrinkwrapFile.ts

```diff
@@ -0,0 +1,95 @@
+import * as fsx from 'fs-extra';
+import * as yaml from 'js-yaml';
+import * as os from 'os';
+
+import { BaseShrinkwrapFile } from '../base/BaseShrinkwrapFile';
+
+interface IShrinkwrapDependencyJson {
+  resolution: {
+    integrity: string;
+    tarball?: string;
+  };
+  dependencies: { [dependency: string]: string };
+}
+
+/**
+ * This interface represents the raw shrinkwrap.YAML file
+ */
+interface IShrinkwrapYaml {
+  /** The list of resolved direct dependencies */
+  dependencies: { [dependency: string]: string };
+/** The description of the solved DAG */
+  packages: { [dependencyVersion: string]: IShrinkwrapDependencyJson };
+  /** URL of the registry */
+  registry: string;
+  shrinkwrapVersion: number;
+  /** The list of specifiers used to resolve direct dependency versions */
+  specifiers: { [dependency: string]: string };
+}
+
+export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
+  private _shrinkwrapJson: IShrinkwrapYaml;
+
+  public static loadFromFile(shrinkwrapYamlFilename: string): PnpmShrinkwrapFile | undefined {
+    try {
+      if (!fsx.existsSync(shrinkwrapYamlFilename)) {
+        return undefined; // file does not exist
+      }
+
+      // We don't use JsonFile/jju here because shrinkwrap.json is a special NPM file format
+      // and typically very large, so we want to load it the same way that NPM does.
+      const parsedData: IShrinkwrapYaml = yaml.safeLoad(fsx.readFileSync(shrinkwrapYamlFilename).toString());
+
+      return new PnpmShrinkwrapFile(parsedData);
+    } catch (error) {
+      throw new Error(`Error reading "${shrinkwrapYamlFilename}":` + os.EOL + `  ${error.message}`);
+    }
+  }
+
+  public getTempProjectNames(): ReadonlyArray<string> {
+    return this._getTempProjectNames(this._shrinkwrapJson.dependencies);
+  }
+
+  // @todo this still has problems!
+  protected getDependencyVersion(dependencyName: string): string | undefined {
+    let dependencyVersion: string | undefined = undefined;
+
+    if (!dependencyVersion) {
+      dependencyVersion = BaseShrinkwrapFile.tryGetValue(this._shrinkwrapJson.dependencies, dependencyName);
+    }
+
+    if (!dependencyVersion) {
+      return undefined;
+    }
+
+    // dependency version can also be a pnpm path such as "/gulp-karma/0.0.5/karma@0.13.22"
+    // in this case we want the first version number that appears. it will be in 3rd spot
+    if (dependencyVersion[0] === '/') {
+      dependencyVersion = dependencyVersion.split('/')[2];
+    }
+
+    return dependencyVersion;
+  }
+
+  private constructor(shrinkwrapJson: IShrinkwrapYaml) {
+    super();
+    this._shrinkwrapJson = shrinkwrapJson;
+
+    // Normalize the data
+    if (!this._shrinkwrapJson.registry) {
+      this._shrinkwrapJson.registry = '';
+    }
+    if (!this._shrinkwrapJson.shrinkwrapVersion) {
+      this._shrinkwrapJson.shrinkwrapVersion = 3; // 3 is the current version for pnpm
```

Comment:
>3 is the current version for pnpm [](start = 53, length = 33)

Why do we need this?

Reply:
I guess we don't technically have to put anything here, but this field is supposed to exist in the yaml file

---
In reply to: [153382126](https://github.com/Microsoft/web-build-tools/pull/449#discussion_r153382126) [](ancestors = 153382126)